build-logic/integration-testing/src/main/kotlin/gradlebuild/integrationtests/buildservice/EmitTeamCityImportDataServiceMessageBuildService.kt

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gradlebuild.integrationtests.buildservice
+
+import org.gradle.api.Project
+import org.gradle.api.provider.Provider
+import org.gradle.api.provider.SetProperty
+import org.gradle.api.services.BuildService
+import org.gradle.api.services.BuildServiceParameters
+import org.gradle.build.event.BuildEventsListenerRegistry
+import org.gradle.kotlin.dsl.support.serviceOf
+import org.gradle.tooling.events.FinishEvent
+import org.gradle.tooling.events.OperationCompletionListener
+import org.gradle.tooling.events.task.TaskFinishEvent
+import org.gradle.tooling.events.task.TaskSuccessResult
+import org.gradle.work.DisableCachingByDefault
+
+
+/**
+ * This is a workaround for https://youtrack.jetbrains.com/issue/TW-76894.
+ *
+ * In short, we want TeamCity to be aware of the test execution data (which tests are executed and how long they are),
+ * even when the Test task is `FROM-CACHE` or `UP-TO-DATE`. This task is registered as a finalizer for the test task
+ * and will output a service message to instruct TeamCity to read JUnit test result XMLs.
+ *
+ * See https://www.jetbrains.com/help/teamcity/service-messages.html#Importing+XML+Reports
+ */
+@DisableCachingByDefault(because = "It outputs a TeamCity service message")
+abstract class EmitTeamCityImportDataServiceMessageBuildService : BuildService<EmitTeamCityImportDataServiceMessageBuildService.Params>, OperationCompletionListener {
+    interface Params : BuildServiceParameters {
+        val includedTaskPaths: SetProperty<String>
+    }
+
+    override fun onFinish(event: FinishEvent) {
+        if (event !is TaskFinishEvent) {
+            return
+        }
+
+        val taskPath = event.descriptor.taskPath
+        if (!parameters.includedTaskPaths.get().contains(taskPath)) {
+            return
+        }
+
+        val taskResult = event.result
+
+        if (taskResult is TaskSuccessResult && (taskResult.isFromCache || taskResult.isUpToDate)) {
+            val (projectName, taskName) = extractProjectAndTask(taskPath)
+            println("##teamcity[importData type='junit' path='subprojects/$projectName/build/test-results/$taskName/TEST-*.xml' verbose='true']")
```

Comment:
What does the `verbose='true'` mean?

Reply:
> verbose='true' attribute will enable detailed logging into the build log.

https://www.jetbrains.com/help/teamcity/service-messages.html#Importing+XML+Reports